iceoryx_utils/test/moduletests/test_concurrent_smart_lock.cpp

```diff
@@ -0,0 +1,358 @@
+{
+    constexpr int32_t CTOR_VALUE = 628189;
+    m_sut.emplace(ForwardArgsToCTor, CTOR_VALUE);
+
+    EXPECT_THAT(m_sut->getCopy().getA(), Eq(CTOR_VALUE));
+}
+
+//////////////////////////////////////
+// END single threaded api test
+//////////////////////////////////////
+
+//////////////////////////////////////
+// BEGIN thread safety tests
+//////////////////////////////////////
+
+// the idea of all tests when the testAction incrementA or constIncrementA
+// would not be thread safe we would encounter either undefined behavior since
+// the variable is sometimes written while it is read or we face a race condition
+// since the increment is always number = number + 1. If this is not thread safe
+// we encounter a jumping in the number and not a steady increment.
+//
+// If no such thing occurs, since the operation is performed in a threadsafe
+// manner, (*m_sut)->getA() == totalSumOfIncrements
+void threadSafeOperationTest(smart_lock_test* test, const std::function<void()> testAction)
+{
+    test->m_sut.emplace(ForwardArgsToCTor, 0);
+
+    vector<std::thread, NUMBER_OF_THREADS> threads;
+    for (uint64_t i = 0U; i < NUMBER_OF_THREADS; ++i)
+    {
+        threads.emplace_back([&] {
+            test->waitUntilThreadsHaveStarted(NUMBER_OF_THREADS);
+            for (uint64_t k = 0U; k < NUMBER_OF_RUNS_PER_THREAD; ++k)
+            {
+                testAction();
+            }
+        });
+    }
+
+    for (uint64_t i = 0U; i < NUMBER_OF_THREADS; ++i)
+    {
+        threads[i].join();
+    }
+}
+
+TEST_F(smart_lock_test, threadSafeAccessThroughArrowOperator)
+{
+    threadSafeOperationTest(this, [=] { (*m_sut)->incrementA(); });
+
+    EXPECT_THAT((*m_sut)->getA(), Eq(NUMBER_OF_RUNS_PER_THREAD * NUMBER_OF_THREADS));
+}
+
+TEST_F(smart_lock_test, threadSafeAccessThroughConstArrowOperator)
+{
+    threadSafeOperationTest(this, [=] { (const_cast<const SutType_t&>(*m_sut))->constIncrementA(); });
+    EXPECT_THAT((*m_sut)->getA(), Eq(NUMBER_OF_RUNS_PER_THREAD * NUMBER_OF_THREADS));
+}
+
+TEST_F(smart_lock_test, threadSafeAccessThroughScopedGuard)
+{
+    threadSafeOperationTest(this, [=] {
+        auto guard = (*m_sut).getScopeGuard();
+        guard->incrementA();
+    });
+    EXPECT_THAT((*m_sut)->getA(), Eq(NUMBER_OF_RUNS_PER_THREAD * NUMBER_OF_THREADS));
+}
+
+TEST_F(smart_lock_test, threadSafeAccessThroughConstScopedGuard)
+{
+    threadSafeOperationTest(this, [=] {
+        auto guard = const_cast<const SutType_t&>(*m_sut).getScopeGuard();
+        guard->incrementA();
+    });
+    EXPECT_THAT((*m_sut)->getA(), Eq(NUMBER_OF_RUNS_PER_THREAD * NUMBER_OF_THREADS));
+}
+
+TEST_F(smart_lock_test, threadSafeCopyCTor)
+{
+    threadSafeOperationTest(this, [=] { SutType_t someCopy(*m_sut); });
+    EXPECT_THAT((*m_sut)->getB(), Eq(NUMBER_OF_RUNS_PER_THREAD * NUMBER_OF_THREADS));
+}
+
+TEST_F(smart_lock_test, threadSafeMoveCTor)
+{
+    threadSafeOperationTest(this, [=] { SutType_t movedSut(std::move(*m_sut)); });
```

Comment:
hmm, you are moving an already moved object ... and still using it later for checks ... bold and reckless ... my inner chicken would probably have picked  a global counter 🐔

Reply:
Actually, to be precise, there is no real moving in C++! I provided an rvalue reference to an assignment operator/constructor and then I can acquire the values of the rvalue - I do not have to invalidate or change it at all. It is up to my implementation!

In here the implementation is clearly defined and we have nowhere undefined behavior or set somewhere some member into an undefined state therefore it is completely legal!

Reply:
oh, this was not a critique but and admiration ... and this is no sarcasm :laughing: